WalletWasabi/CoinJoin/Client/Clients/CoinJoinClient.cs

```diff
@@ -685,6 +686,19 @@ private async Task TryRegisterCoinsAsync(ClientRound inputRegistrableRound)
 			return (change, actives);
 		}
 
+		private void CleanNonLockedExposedKeys()
+		{
+			// Remove non-locked exposed keys.
+			foreach (var key in ExposedLinks.Keys.ToArray())
+			{
+				ExposedLinks[key] = ExposedLinks[key].Where(x => x.Key.KeyState == KeyState.Locked).ToArray();
+				if (!ExposedLinks[key].Any())
+				{
+					ExposedLinks.TryRemove(key, out _);
+				}
```

Comment:
AddOrReplace is better because it is threadsafe.

```suggestion
				var lockedKeys = ExposedLinks[key].Where(x => x.Key.KeyState == KeyState.Locked).ToArray();
				if (lockedKeys.Any())
				{
					ExposedLinks.AddOrReplace(key, lockedKeys);
				}
				else
				{
					ExposedLinks.TryRemove(key, out _);
				}
```